File: algorithms/20 - Valid Parentheses.cpp
//Description of the problem - https://leetcode.com/problems/valid-parentheses/

class Solution {
public:
    bool isValid(string s) {
        stack<char> checking;
        int stringLength = s.size();
        
        for(int i = 0; i < stringLength; i++){
            char currentChar = s[i];
            
            if(currentChar == '(' || currentChar == '{' || currentChar == '[') {
                checking.push(currentChar);
            }
            else if (checking.empty()) {
                return false;
            }
            else if(currentChar == ')') {
                if (checking.top() != '(') return false;
                checking.pop();
            } 
            else if(currentChar == '}') {
                if (checking.top() != '{') return false;
                checking.pop();
            } 
            else if(currentChar == ']') {
                if (checking.top() != '[') return false;
                checking.pop();
            }
        }
        
        if(checking.empty()) return true;
        return false;
    }
};

File: algorithms/20 - Valid Parentheses.go
func isValid(s string) bool {
    
    var stack []byte
    
    for i := 0; i < len(s); i++ {
        var char byte = s[i]
        
        if (len(stack) != 0) {
            top := stack[len(stack)-1]
            
            if (char == '}' && top == '{') {
                stack = stack[:len(stack)-1] 
                continue
            } else if (char == ']' && top == '[') {
                stack = stack[:len(stack)-1] 
                continue
            } else if (char == ')' && top == '(') {
                stack = stack[:len(stack)-1]
                continue
            }   
        }
        stack = append(stack, char)   
    } 
    
    if (len(stack) != 0) { 
        return false
    }
    
    return true
}
 